File: 118. Pascal's Triangle/main.py
"""
Author: Amal Matty Antony
"""
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        base = [[1],[1,1]]
        if numRows <= 2:
            return base[:numRows]
        for row in range(3,numRows+1):
            new = [1]
            prev_row = base[row-2]
            for i in range(row-2):
                new.append(prev_row[i]+prev_row[i+1])
            new.append(1)
            base.append(new)
        return base
            

File: 118. Pascal's Triangle/main.cpp
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> solution;
        solution.push_back(vector<int>{1});
        if(numRows == 1) return solution;
        vector<int> cache = {1,1};
        solution.push_back(cache);
        for(int i = 2;i<numRows;i++){
            vector<int> newRow(i+1,1);
            for(int j=1;j<=i/2;j++){
                newRow[j] = newRow[i-j] = cache[j-1] + cache[j];
            }
            solution.push_back(newRow);
            cache = newRow;
        }
        return solution;
    }
};